# Task-1: Data Quality Issues _ CustomerDemographic

The client has provided us with four datasets.
They are:
1) Customer Demographic
2) Transaction details
3) Customer address
4) New customer list

In [1]:
#Import the required librariesabs
import pandas as pd
import numpy as np

## Reading Excel file given by the client

In [2]:
df=pd.read_excel("G:\My Drive\KPMG\Task 1\KPMG_ImportPython.xlsx","CustomerDemographic")

Let us copy the same data frame (as below) so that changes to be made in future may not affect the Excel file

In [3]:
df=df.copy()

## Viewing the data frame

In [4]:
df.head(5)

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

## Exploring shape of the sheet 

In [5]:
df.shape

(4000, 13)

## Exploring columns of the sheets 

In [6]:
df.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

## Checking Data Types of each attribute in each sheet

In [7]:
df.dtypes

customer_id                                     int64
first_name                                     object
last_name                                      object
gender                                         object
past_3_years_bike_related_purchases             int64
DOB                                    datetime64[ns]
job_title                                      object
job_industry_category                          object
wealth_segment                                 object
deceased_indicator                             object
default                                        object
owns_car                                       object
tenure                                        float64
dtype: object

## Describe the data columnwise use "df.describe()"

## Checking for Null/missing values

In [8]:
df.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

## Exploring columns 

In [9]:
df.iloc[:,7].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

Here, in [], : before comma indicates the row index, here : means all rows. Similarly values after comma indicates the column index, here 7 means 8th column Because, the indices in python start with zero.

Below is the other way to explore the same column, i.e. Transactions.iloc[:,7] = job_industry_category

In [10]:
df['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

## Why to see value_counts seperately for each column?, Make a loop so that together we get the desired answer for all columns.. 

In [11]:
df.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

In [12]:
df_columns=['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure']

In [13]:
for i in df_columns:
    print(df[i].value_counts())

2047    1
657     1
2732    1
681     1
2728    1
677     1
2724    1
673     1
2720    1
669     1
2716    1
665     1
2712    1
661     1
2708    1
2704    1
625     1
653     1
2700    1
649     1
2696    1
645     1
2692    1
641     1
2688    1
637     1
2684    1
633     1
2680    1
629     1
       ..
3419    1
3363    1
1370    1
3415    1
1366    1
3411    1
1362    1
3407    1
1358    1
3403    1
1354    1
3399    1
1350    1
3395    1
1346    1
3391    1
1342    1
3387    1
1338    1
3383    1
1334    1
3379    1
1330    1
3375    1
1326    1
3371    1
1322    1
3367    1
1318    1
2049    1
Name: customer_id, Length: 4000, dtype: int64
Max            5
Timmie         5
Tobe           5
Kippy          4
Pail           4
Basile         4
Urbanus        4
Gerome         4
Emlyn          4
Lindsay        4
Osgood         4
Corabelle      4
Kim            4
Hubie          4
Andrea         3
Andreas        3
Normand        3
Jayme          3
Erik           3
Lauree         3
Coop

## Determining percetage of  missing values in each column.

In [14]:
missing_count=df.isnull().sum()
missing_count

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [15]:
total=len(df)
total

4000

In [16]:
percentageMissing=missing_count*100/total
percentageMissing

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
default                                 7.550
owns_car                                0.000
tenure                                  2.175
dtype: float64

In [17]:
missing_data=pd.DataFrame({'Heading': df.columns, 'total_missing': missing_count,'Percentage_missing': percentageMissing})
missing_data.sort_values('total_missing', ascending=False)
missing_data=missing_data.loc[missing_data['total_missing'] > 0]
missing_data

Heading  total_missing  \
last_name                          last_name            125   
DOB                                      DOB             87   
job_title                          job_title            506   
job_industry_category  job_industry_category            656   
default                              default            302   
tenure                                tenure             87   

                       Percentage_missing  
last_name                           3.125  
DOB                                 2.175  
job_title                          12.650  
job_industry_category              16.400  
default                             7.550  
tenure                              2.175

In [18]:
missing_data.to_csv('missing_data_CustomerDemographic.csv', index=False)

## Deteriming if duplicate value is there in the data frame

In [19]:
duplicate=df[df.duplicated()]
duplicate

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, default, owns_car, tenure]
Index: []

In [20]:
len(duplicate)

0

## Viewing categorical data

In [21]:
df['gender'].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

Ohhh !! There is inconsistency in data. This issue in column 'gender' needs to be addressed. Before that let us check another columns..

In [22]:
df['job_title'].value_counts()

Business Systems Development Analyst    45
Social Worker                           44
Tax Accountant                          44
Internal Auditor                        42
Legal Assistant                         41
Recruiting Manager                      41
General Manager                         40
Associate Professor                     39
Structural Engineer                     39
Junior Executive                        38
Desktop Support Technician              38
Sales Representative                    38
Senior Financial Analyst                37
Technical Writer                        37
Executive Secretary                     37
Chemical Engineer                       37
Dental Hygienist                        37
Registered Nurse                        36
Senior Developer                        36
Cost Accountant                         36
Assistant Professor                     36
Product Engineer                        36
Food Chemist                            35
VP Product 

In [23]:
df['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [24]:
df['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [25]:
df['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [26]:
df['owns_car'].value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

## Solving issues of 'gender' column's inconsistency 

In [27]:
df=df.replace({'gender':{'Female':'F','Male':'M','Femal':'F'}})
df['gender'].value_counts()

F    2039
M    1873
U      88
Name: gender, dtype: int64

## Adding Age (Calculated from DOB) into data frame 

In [28]:
now = pd.Timestamp('now')

In [29]:
df['DOB']=pd.to_datetime(df['DOB'], format='%m%d%y')
df['DOB'].head()

0   1953-10-12
1   1980-12-16
2   1954-01-20
3   1961-10-03
4   1977-05-13
Name: DOB, dtype: datetime64[ns]

In [30]:
df['DOB']=df['DOB'].where(df['DOB']<now,df['DOB']-np.timedelta64(100,'Y'))
df['age']=(now-df['DOB']).astype('<m8[Y]')
df['age'].head()

0    69.0
1    41.0
2    68.0
3    61.0
4    45.0
Name: age, dtype: float64

In [32]:
df['age'].describe()

count    3913.000000
mean       44.885254
std        12.791863
min        20.000000
25%        35.000000
50%        45.000000
75%        54.000000
max       178.000000
Name: age, dtype: float64

Maximum age observed is 178 years. There may be typing mistake. Let us remove any record age of which is beyond 100 years.

In [37]:
indices = df[df['age']>100 ].index
indices

Int64Index([33], dtype='int64')

In [38]:
df.drop(indices, inplace = True)

## Save the data frame in '.csv' fromat

In [31]:
df.to_csv('CustomerDemographic_Done.csv', index=False)